diff --git a/part/syntax/katehighlight.cpp b/part/syntax/katehighlight.cpp
index 7d956e5..cad75b0 100644
--- a/part/syntax/katehighlight.cpp
+++ b/part/syntax/katehighlight.cpp
@@ -192,19 +192,6 @@ KateHlContext *KateHighlighting::generateContextStack (Kate::TextLineData::Conte
         {
           // set new index, if stack is empty, this is -1, done for eternity...
           indexLastContextPreviousLine = contextStack.size() - 1;
-
-          // stack already empty, nothing to do...
-          if ( contextStack.isEmpty() )
-            return contextNum (0);
-
-          KateHlContext *c = contextNum(contextStack.last());
-
-          // this must be a valid context, or our context stack is borked....
-          Q_ASSERT (c);
-
-          // handle line end context as new modificationContext
-          modification = c->lineEndContext;
-          continue;
         }
 
         return contextNum (contextStack.isEmpty() ? 0 : contextStack.last());
